pf/tfbridge/ignore_changes.go

```diff
@@ -0,0 +1,301 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tfbridge
+
+import (
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+
+	"github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+
+	"github.com/pulumi/pulumi-terraform-bridge/pf/internal/convert"
+)
+
+// Implements support for ignoreChanges property.
+//
+// There is some complexity here since Plugin Framework returns diffs in terms of Terraform paths, so the implementation
+// tries to translate TF paths to Pulumi paths to decide if they match.
+//
+// See also https://www.pulumi.com/docs/intro/concepts/resources/options/ignorechanges/
+type ignoreChanges struct {
+	// Delegate syntax parsing to resource.PropertyPath.
+	paths []resource.PropertyPath
+	ss    schemaStepper
+}
+
+func newIgnoreChanges(
+	schema *schema.PackageSpec,
+	token tokens.Token,
+	renames convert.PropertyNames,
+	rawIgnoreChanges []string,
+) (*ignoreChanges, error) {
+	var paths []resource.PropertyPath
+	for _, p := range rawIgnoreChanges {
+		pp, err := resource.ParsePropertyPath(p)
+		if err != nil {
+			return nil, err
+		}
+		paths = append(paths, pp)
+	}
+	return &ignoreChanges{
+		paths: paths,
+		ss: &namedEntitySchemaStepper{
+			schema: schema,
+			token:  token,
+			// Another complication is that Pulumi renames the properties, this uses the Renames framework
+			// to track the name tables.
+			renames: renames,
+		},
+	}, nil
+}
+
+func (ic *ignoreChanges) IsIgnored(ap *tftypes.AttributePath) bool {
+	for _, p := range ic.paths {
+		if ic.isIgnoredBy(p, ap) {
+			return true
+		}
+	}
+	return false
+}
+
+func (ic *ignoreChanges) isIgnoredBy(pattern resource.PropertyPath, ap *tftypes.AttributePath) bool {
+	mc := &matchingContext{
+		remainingPattern: pattern,
+		schemaStepper:    ic.ss,
+	}
+	match, remain, err := tftypes.WalkAttributePath(mc, ap)
+	if err != nil {
+		return false
+	}
+	if remain != nil && len(remain.Steps()) > 0 {
+		return false
+	}
+	if m, ok := match.(*matchingContext); ok {
+		if !m.matchFailed {
+			return true
+		}
+	}
+	return false
+}
+
+type schemaStepper interface {
+	Property(resource.PropertyKey) schemaStepper
+	Element() schemaStepper
+	PropertyKey(convert.TerraformPropertyName) resource.PropertyKey
+}
+
+type typeSchemaStepper struct {
```

Comment:
Can I get a doc comment explaining what this is? I read it as "The type of a schema stepper", but that seems wrong.

Reply:
I can try, but these types should have been anonymous.. This is all machinery to support matching algo reusing tftypes.WalkAttributePath...